donut/modules/calendar/static/script.js

```diff
@@ -0,0 +1,388 @@
+{
+ data.events = data.events || [];
+
+  for(var i = 0; i < data.events.length; i++)
+  {
+    var curEvent = data.events[i];
+    var beginStr = curEvent.start.dateTime || curEvent.start.date;
+    var beginMonth = parseInt(beginStr.substring(5, 7));
+    var endStr = curEvent.end.dateTime || curEvent.end.date;
+    var endMonth = parseInt(endStr.substring(5, 7));
+
+    // If the current year and the end year are not the same, then 
+    // for whatever ungodly reason the event spans a year
+    endMonth = (new Date(endStr).getFullYear() === year) ? endMonth : 12;
+
+    // Same but with beginning months
+    beginMonth =  (new Date(beginStr).getFullYear() === year) ? beginMonth : 1;
+    while(beginMonth <= endMonth)
+    {
+      events[beginMonth-1].push(curEvent);
+      beginMonth += 1;
+    }
+  }
+  renderEvents(month, year);
+}
+
+// Actually requests for the events from the server. 
+function getEvents(month, year){
+  // We don't always need to send a request. 
+  if (month === 11||month === 0 || $.isEmptyObject(events) || curYear != year)
+  {
+    curYear = year;
+    $('div').filter($('.day')).each(function(){$(this).text('loading')});
+    for (var i = 0; i < 12; i++)
+    {
+        events[i] = [];
+    }
+    $.ajax({
+      url: '/1/calendar_events',
+      type: 'POST',
+      data:{
+        month:month+1, 
+        year:year
+      },
+      success:function(res){
+          process_events(res, month, year);
+          $('#failover_message').hide();
+      },
+      // Perhaps the timeout is too long... But this is about the time where
+      // Half of the time it'll work and the other half wouldn't
+      timeout:3000, 
+      error: function(jqXHR, textStatus){
+        if(textStatus === 'timeout')
+        {     
+            // Access the backup
+            $.ajax({
+                url: '/1/calendar_events_backup',
+                type: 'POST',
+                data:{
+                    month:month+1,
+                    year:year
+                },
+                success:function(data){
+                    if (!window.parent.location.href.includes('sync')){
+                        $('#failover_message').show();
+                    }
+                    process_events(data, month, year); 
+                }
+            });
+        }
+      }
+    });
+
+  }
+}
+
+// Render Calendar
+function renderCalendar(startDay, totalDays, currentDate) {
+  var currentRow = 1;
+  var currentDay = startDay;
+  var $table = $('#cal');
+  var $week = getCalendarRow();
+  var $day;
+  var i = 1;
+  for (; i <= totalDays; i++) {
+    $day = $week.find('td').eq(currentDay);
+    $day.text(i);
+    if (i === currentDate) {
+      $day.addClass('today');
+    }
+
+    $day.append("</br>");
+    var wrapper = $('<div></div>').attr('id', 'day '+i).addClass('day');
+    $day.append(wrapper);
+    // +1 next day until Saturday (6), then reset to Sunday (0)
+    currentDay = ++currentDay % 7;
```

Comment:
It might be cleaner to do the iteration by incrementing a `Date` object calling `getDay()` on it